File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/1250-LongestCommonSubsequence/1250-LongestCommonSubsequence.cpp
// Last updated: 8/4/2026, 7:09:06 PM
class Solution {
public:
    int dp[1001][1001];
    int longestCommonSubsequence(string text1, string text2) {
        for(int i = 0 ; i < 1001 ; i++){
            for(int j = 0 ; j < 1001 ; j++){
                dp[i][j] = -1;
            }
        }
        return lcs(text1 , text2 , text1.length() , text2.length());
    }
    int lcs(string &x , string &y , int n , int m){
        if(n == 0 || m == 0){
            return 0;
        }
        if(dp[n][m] != -1) return dp[n][m];
        if(x[n-1] == y[m-1]){
            return dp[n][m] = 1 + lcs(x , y , n-1 , m-1);
        }
        else{
            return dp[n][m] = max(lcs(x , y , n-1 , m) , lcs(x , y , n , m-1));
        }
    }
};

File: https:/github.com/abhinavtripathi17/DSA-Leetcode.git/1250-LongestCommonSubsequence/1250-LongestCommonSubsequence.java
// Last updated: 8/4/2026, 7:09:13 PM
class Solution {
    int dp[][] = new int[1001][1001];
    public int longestCommonSubsequence(String text1, String text2) {
        int n = text1.length();
        int m = text2.length();

        for(int i = 1 ; i <= n ; i++){
            for(int j = 1 ; j <= m ; j++){
                if(text1.charAt(i-1) == text2.charAt(j-1)){
                    dp[i][j] = 1 + dp[i-1][j-1];
                }
                else{
                    dp[i][j] = Math.max(dp[i-1][j] , dp[i][j-1]);
                }
            }
        }

        return dp[n][m];
    }


}